appscale/tools/agents/azure_agent.py

```diff
@@ -0,0 +1,689 @@
+  AZURE_AUTH_ENDPOINT = 'https://login.microsoftonline.com/'
+
+  # The Azure Resource URL to get the authentication token using client credentials.
+  AZURE_RESOURCE_URL = 'https://management.core.windows.net/'
+
+  # Default Storage Account name to use for Azure.
+  DEFAULT_STORAGE_ACCT = 'appscalestorage'
+
+  # Default resource group name to use for Azure.
+  DEFAULT_RESOURCE_GROUP = 'appscalegroup'
+
+  # The following constants are string literals that can be used by callers to
+  # index into the parameters that the user passes in, as opposed to having to
+  # type out the strings each time we need them.
+  PARAM_APP_ID = 'app_id'
+  PARAM_APP_SECRET = 'app_secret_key'
+  PARAM_CREDENTIALS = 'credentials'
+  PARAM_EXISTING_RG = 'does_exist'
+  PARAM_GROUP = 'group'
+  PARAM_INSTANCE_IDS = 'instance_ids'
+  PARAM_KEYNAME = 'keyname'
+  PARAM_IMAGE_ID = 'image_id'
+  PARAM_REGION = 'region'
+  PARAM_RESOURCE_GROUP = 'resource_group'
+  PARAM_STORAGE_ACCOUNT = 'storage_account'
+  PARAM_SUBCR_ID = 'subscription_id'
+  PARAM_TENANT_ID = 'tenant_id'
+  PARAM_TEST = 'test'
+  PARAM_TAG = 'group_tag'
+  PARAM_VERBOSE = 'is_verbose'
+  PARAM_ZONE = 'zone'
+
+  # A set that contains all of the items necessary to run AppScale in Azure.
+  REQUIRED_CREDENTIALS = (
+    PARAM_APP_SECRET,
+    PARAM_APP_ID,
+    PARAM_IMAGE_ID,
+    PARAM_KEYNAME,
+    PARAM_SUBCR_ID,
+    PARAM_TENANT_ID,
+    PARAM_ZONE
+  )
+
+  # The admin username needed to create an Azure VM instance.
+  ADMIN_USERNAME = 'azureuser'
+
+  # The maximum number of seconds to sleep while waiting for
+  # Azure resources to get created.
+  SLEEP_TIME = 10
+
+  def assert_credentials_are_valid(self, parameters):
+    """ Contacts Azure with the given credentials to ensure that they are
+    valid. Gets an access token and a Credentials instance in order to be
+    able to access any resources.
+    Args:
+      parameters: A dict, containing all the parameters necessary to
+        authenticate this user with Azure.
+    Returns:
+      True, if the credentials were valid.
+      A list, of resource group names under the subscription.
+    Raises:
+      AgentConfigurationException: If an error is encountered during
+        authentication.
+    """
+    credentials = self.open_connection(parameters)
+    subscription_id = parameters[self.PARAM_SUBCR_ID]
+    try:
+      resource_client = ResourceManagementClient(credentials, subscription_id)
+      resource_groups = resource_client.resource_groups.list()
+      rg_names = []
+      for rg in resource_groups:
+        rg_names.append(rg.name)
+      return True, rg_names
+    except CloudError as error:
+      raise AgentConfigurationException("Unable to authenticate using the "
+        "credentials provided. Reason: {}".format(error.message))
+
+  def configure_instance_security(self, parameters):
+    """Configure and setup groups and storage accounts for the VMs spawned.
```

Comment:
Are storage accounts actually being initiated here? If so, it doesn't make much sense logically given the name of the method.